### PR TITLE
New scheduler

### PR DESCRIPTION
**Pull request by Soroosh129**
This adds a modular API for runtime scheduler development for the threaded C runtime.

It also adds the following tested schedulers as a starting point:
- NP: A non-preemptive scheduler
- GEDF_NP: A Global Earliest Deadline First non-preemptive scheduler
- GEDF_NP_CI: A Global Earliest Deadline First non-preemptive scheduler with chain ID

These schedulers are tested by running all the LF tests for them (wherever they are supported).

This PR also contains the following experimental (untested) scheduler:
- PEDF_NP: Partitioned Earliest Deadline First non-preemptive scheduler.

Possible future directions include:
- Adding unit tests for schedulers (how could we meaningfully test these schedulers?)
- Further developing the PEDF_NP scheduler
- Developing preemptive versions of the existing schedulers
- Developing new schedulers (for example, the ones mentioned in Jane Liu's Real-Time Systems book) such as RM
- Developing new methods to lift the global synchronization barrier on the advancement of tag.

Here are the results of using the new schedulers for the Savina benchmark suite compared to the scheduler in `main` and the C++ runtime. These results are obtained by @petervdonovan (thanks Peter a ton!) on ~~December 22nd~~ February 2nd:

e76957c34-benchmarks

Oddities to mention include the fact that the default NP scheduler performs much worse than others for Cigarette Smokers, Concurrent Sorted List, Radix Sort and Sleeping Barber. Interestingly, the only difference between NP and GEDF_NP is that the NP scheduler uses a `vector` that is accessed by an index that is decremented via an atomic operation, whereas GEDF_NP uses a `pqueue` that is accessed using push and pop mechanisms with mutex locks. This pattern seems to pay off in NP's favor for benchmarks such as Fork Join (throughput), Counting Actor, Thread Ring and Ping Pong, but of course, it negatively affects some others.

### CODE REVIEW

**Comment by lhstrh** [comment]
Note: the weird code coverage stuff that's happening is due to https://github.com/lf-lang/lingua-franca/issues/836

File: core/threaded/scheduler.h
Comment (edwardalee) [inline core/threaded/scheduler.h:27]: I'm guessing that this file defines a set of functions that have to implemented by a particular scheduler, right?  I comment here to this effect would be helpful.  

The number of functions is surprisingly small, which is a very good sign about the design.

**Review by edwardalee — APPROVED** [review]
This is pretty large amount of work and I left some very minor comments, but overall, this looks extremely well organized and well documented to me.  I think if all tests are passing with each of the schedulers, this can be merged. Do we know what the performance impact is?

**Review by petervdonovan — APPROVED** [review]
I left a few questions and comments. Not all of them necessarily should be addressed before this is merged -- they are just what occurred to me as I read these diffs.

File: core/threaded/scheduler_GEDF_NP.c
Comment (Soroosh129) [reply core/threaded/scheduler_GEDF_NP.c:58]: > One thread going idle can change _lf_sched_semaphore->count by more than one.

Yes, the comment on the semaphore is incorrect. The semaphore doesn't actually keep track of the number of idle worker threads. The semaphore will be released by the scheduler with the count being the _maximum_ number of worker threads that can wake up (with the optimization that the calling worker thread doesn't need a notification, so it would be excluded from the semaphore count). If one worker thread is fast enough to go through all the reactions before other worker threads can wake up, that is not an error condition. This is a best-effort method of work distribution. In this case, a greedy worker is not necessarily a problem. This greediness is inherent in the fact that all worker threads share the same work queue. A fast worker thread can go through all the work before others get the chance.

On the other hand, the PEDF scheduler has a more direct method of waking up specific worker threads because each worker thread has its own work queue (although it does have work-stealing if it runs out of work).


> It looks like the semaphore is not providing more functionality than the lf_cond_t that it uses, because it ensures that workers block when they finish and the reaction queue is empty, and it ensures that they unblock when _lf_sched_notify_workers is invoked.

This is not true. The semaphore is in charge of controlling the _maximum_ (not exact) number of worker threads that can wake up. This would not be trivially possible using a `lf_cond_t`.

File: core/threaded/scheduler.h
Comment (Soroosh129) [reply core/threaded/scheduler.h:59]: Good point.

File: core/threaded/scheduler_GEDF_NP.c
Comment (Soroosh129) [reply core/threaded/scheduler_GEDF_NP.c:76]: Maybe. I think it would have to be `NUMBER_OF_REACTION_LEVELS` I think to clarify what levels we are talking about. The +1 here is just a nuisance.

File: core/threaded/scheduler_GEDF_NP.c
Comment (Soroosh129) [reply core/threaded/scheduler_GEDF_NP.c:162]: Yes.

File: core/threaded/scheduler_GEDF_NP.c
Comment (Soroosh129) [reply core/threaded/scheduler_GEDF_NP.c:192]: `assert` (or any check) can be expensive, that's why it is commented out entirely. In C++, having a `Release` build type disables asserts, but I'm not sure if it works for the C `assert` macro.
Moreover, `executing_q` is local to the scheduler, so this would not qualify as a precondition. 
We can bite the bullet and add the check.

File: core/utils/pqueue_support.h
Comment (Soroosh129) [reply core/utils/pqueue_support.h:54]: Ah, this is a relic of the past. Good catch!

File: core/utils/semaphore.c
Comment (Soroosh129) [reply core/utils/semaphore.c:57]: You're right about the use case. But the return value is not checked in the schedulers for the sake of performance. I still think that the correct C API design for the semaphore is for it to return a status code. This semaphore can be useful in other parts of reactor-c too.

File: core/utils/semaphore.c
Comment (Soroosh129) [reply core/utils/semaphore.c:61]: Nice catch! Thanks.

File: core/federated/RTI/rti.c
Comment (Soroosh129) [reply core/federated/RTI/rti.c:62]: Whoops, outdated comment.

File: core/reactor.c
Comment (Soroosh129) [reply core/reactor.c:396]: Yes, it is. This line of code would result in runtime exceptions complaining about double freeing a malloced memory. I would have to dig deeper into this issue.

File: core/threaded/scheduler.h
Comment (Soroosh129) [reply core/threaded/scheduler.h:27]: > I'm guessing that this file defines a set of functions that have to implemented by a particular scheduler, right? I comment here to this effect would be helpful.

Yes, this makes sense. I will add these.

File: core/threaded/scheduler_GEDF_NP_CI.c
Comment (Soroosh129) [reply core/threaded/scheduler_GEDF_NP_CI.c:94]: I agree. I did some preliminary benchmarks and settled on a `pqueue_t`, but it should be just as easy to switch to something like a vector.

File: core/threaded/scheduler_GEDF_NP.c
Comment (petervdonovan) [reply core/threaded/scheduler_GEDF_NP.c:192]: Empirically, it has not appeared to me that `assert` had any overhead outside of debug mode. [This post](https://stackoverflow.com/questions/1571340/what-is-the-assert-function) says it shouldn't, and when I look in `/usr/include/assert.h`, I see handling of `NDEBUG` that looks right, although I don't understand everything that happens.

File: core/threaded/scheduler_GEDF_NP_CI.c
Comment (Soroosh129) [reply core/threaded/scheduler_GEDF_NP_CI.c:513]: Good catch! Thanks.

File: core/threaded/scheduler_NP.c
Comment (Soroosh129) [reply core/threaded/scheduler_NP.c:96]: I dabbled with an atomic vector implementation for a bit, but in the end found it to need a lock when it is expanding, so I gave up on it. If you would like to take a stab at it, I might have a half-finished version somewhere.

File: core/threaded/scheduler_GEDF_NP_CI.c
Comment (petervdonovan) [reply core/threaded/scheduler_GEDF_NP_CI.c:94]: Thanks for explaining -- the truth is surprising to me, but it resolves my comment.

File: core/threaded/scheduler_NP.c
Comment (Soroosh129) [reply core/threaded/scheduler_NP.c:130]: This is a great point!

File: core/threaded/scheduler_GEDF_NP.c
Comment (Soroosh129) [reply core/threaded/scheduler_GEDF_NP.c:192]: Okay, that's promising. Then we can safely go with an assert here.

File: core/utils/semaphore.c
Comment (petervdonovan) [reply core/utils/semaphore.c:57]: I do not disagree with your general point. What I suggested was not very safe. But there were reasons why I thought an `assert` might be a fair alternative here:
* If performance matters in the schedulers, then it may matter here
* This is different from functions like `malloc` (which actually might fail, through no fault of the caller) in that if this check fails, then it was because of an illegal argument -- the caller shot themselves in the foot. If the caller is not responsible enough to ensure that the pointer is not NULL, then why should the caller be responsible enough to check the return code?
* In many cases, not doing the check when the semaphore is NULL would result in a segfault (I think?), which would crash the program and allow the user to know that something is terribly wrong. In this sense, this behavior could even be more idiot-proof in some ways than relying on the user to check the return code.

File: core/threaded/scheduler_NP.c
Comment (petervdonovan) [reply core/threaded/scheduler_NP.c:96]: Oh, that is a good point -- it sounds like it would have to either never resize or rely on a lock instead of atomic operations. I would be happy to take that as an action item, but my instinct is that it is not a good idea to let that get in the way of this PR.

File: core/utils/semaphore.c
Comment (Soroosh129) [reply core/utils/semaphore.c:57]: Good point. Let's use your simpler API design.

File: core/threaded/scheduler_GEDF_NP.c
Comment (petervdonovan) [reply core/threaded/scheduler_GEDF_NP.c:58]: Thanks for your explanation! I'll have to re-read the code now that I know what the semaphore signifies.

**Comment by petervdonovan** [comment]
Sorry for the CodeCov report. I think it's once again because of changes to the workflow in the `lingua-franca` repo. I will submit a patch.

File: core/threaded/scheduler_NP.c
Comment (Soroosh129) [reply core/threaded/scheduler_NP.c:96]: There is an optimization that could make the vector access lock-free, at least for this scheduler. If we knew beforehand the maximum number of reactions per level, we could just statically pre-allocate the necessary memory. After talking to @edwardalee about it, we concluded that it would make the most sense to do this after the scalability-banks branch is merged in.

**Comment by Soroosh129** [comment]
> Sorry for the CodeCov report. I think it's once again because of changes to the workflow in the `lingua-franca` repo. I will submit a patch.

No worries, I just deleted it for now.

File: core/threaded/scheduler_GEDF_NP.c
Comment (Soroosh129) [reply core/threaded/scheduler_GEDF_NP.c]: I updated the comment to better reflect this point.

File: core/threaded/scheduler_GEDF_NP_CI.c
Comment (Soroosh129) [reply core/threaded/scheduler_GEDF_NP_CI.c:140]: Funny enough, I think the atomic operation is seemingly faster than directly setting the variable on PC. I agree that it shouldn't make a difference (and not having it as atomic makes more sense).

File: core/threaded/scheduler_GEDF_NP_CI.c
Comment (Soroosh129) [reply core/threaded/scheduler_GEDF_NP_CI.c:184]: I think I don't understand your point.

For this particular optimization (introduced in #17), the idea is that if the entire index of the `reaction` at hand is already smaller than the index at the head of the `executing_q`,  there is nothing that can block the `reaction` from executing. The executing_q is sorted such that the smallest deadline and level end up at the top.

Are you suggesting that only the level should be checked?

Comment (petervdonovan) [reply core/threaded/scheduler_GEDF_NP_CI.c:184]: What if the first reaction in the executing queue has a later deadline but an earlier level than `reaction`? To me, it looked like this expression would evaluate to `true` in that case because the deadline is in a more significant bit position, and the entire function call would return `false`. But it should return `true`, because the reaction in the executing queue that is at an earlier level might be blocking `reaction`.

I just brought this up to be extra sure, since I'm not sure if we have a lot of test cases involving multiple reactions with different deadlines.

File: core/threaded/scheduler.h
Comment (Soroosh129) [inline core/threaded/scheduler.h:32]:
```suggestion
 * implementation for functions that are defined in this header file.
```

File: core/threaded/scheduler.h
Comment (Soroosh129) [inline core/threaded/scheduler.h:44]:
```suggestion
 * @brief Default value that is assumed to be the maximum reaction level in the
```

File: core/threaded/scheduler.h
Comment (Soroosh129) [inline core/threaded/scheduler.h:102]:
```suggestion
 * @param worker_number For the calling worker thread.
```

**Comment by Soroosh129** [comment]
Any objections to merge this?

There is already a merge conflict since the last time I resolved conflicts, which was 10 days ago.

If it's okay with everyone, I will create an issue for the testing and merge this.

**Comment by edwardalee** [comment]
OK with me to merge.

**Comment by lhstrh** [comment]
I would really prefer to get the testing done first. I will get it done this weekend, like I said earlier today.

**Comment by lhstrh** [comment]
Also, the conflict is really minor. So there's no need to ring the alarm bells about this.

**Comment by Soroosh129** [comment]
Okay, if you think fixing the tests is critical to merge this into reactor-c, let's wait. I did create an issue (#39).

**Comment by a-sr** [comment]
> It looks like `_lf_handle_mode_changes();` can be moved to `_lf_next_locked();` and `_lf_mode_is_active` can be checked in `_lf_trigger_reaction` (the latter change also needs to happen for `core/reactor.c`). I think the good news seems to be that nothing needs to move to the new schedulers.

`_lf_handle_mode_changes` must be called once after all reactions at a given time tag were processed but before the next tag is determined because it may change the event queue. The current location in `_lf_worker_advance_tag_locked` was just a guess where this condition might hold and it worked. I did not put it in `_lf_next_locked` because its purpose seemed solely in advancing time and it is surrounded by `tracepoint_worker_advancing_time_*`, so I did not want to meddle with this. But if you say it is ok to put it in `_lf_next_locked`, that is fine by me.

`_lf_trigger_reaction` looks like it should work the same in preventing reactions that are in inactive modes from entering the reaction queue.

I also suppress some reactions, triggers, and timers in `_lf_pop_events`, `_lf_initialize_timer`, and `_lf_insert_reactions_for_trigger` of `reactor_common.c`. Will this also work in threaded?

Otherwise, I do not see any potential conflicts/problems.

**Comment by Soroosh129** [comment]
Thanks @a-sr for your explanation.

The functionality of `_lf_worker_advance_tag_locked` is no longer in `reactor_threaded.c` in this PR. The job of advancing tag is delegated to the schedulers and might be scheduler-specific. `_lf_next_locked` is for now used by all the schedulers, but this could change if a scheduler is created that does not use the current synchronization barrier on advancing tag. I don't think we should worry about this now though. That's why I suggested to put `_lf_handle_mode_changes` in `_lf_next_locked`. Another place for `_lf_handle_mode_changes` could be in [`scheduler_sync_tag_advance.c`](https://github.com/lf-lang/reactor-c/blob/8778cec7f7c5f12283769acccd7b9bc1f35e52c6/core/threaded/scheduler_sync_tag_advance.c).

> _lf_trigger_reaction looks like it should work the same in preventing reactions that are in inactive modes from entering the reaction queue.

I agree. This would eliminate the need to repeat `_lf_mode_is_active` in each scheduler.

> I also suppress some reactions, triggers, and timers in _lf_pop_events, _lf_initialize_timer, and _lf_insert_reactions_for_trigger of reactor_common.c. Will this also work in threaded?

I believe this should be fine. `reactor_common.c` hasn't changed much in this PR and I don't immediately see a red flag in terms of conflicts.

**Comment by lhstrh** [comment]
@a-sr, would it be OK to merge this in once I complete my review, or is there any reason to hold off?

**Comment by a-sr** [comment]
Yes, it can be merged, the effort to resolve conflicts on the modal models branch is definitely manageable.

File: core/threaded/scheduler_GEDF_NP_CI.c
Comment (lhstrh) [inline core/threaded/scheduler_GEDF_NP_CI.c:439]: This, too, we should document in an issue.

File: core/threaded/scheduler_GEDF_NP_CI.c
Comment (lhstrh) [inline core/threaded/scheduler_GEDF_NP_CI.c:406]: I know this was inherited, but I'm a bit confused about this now. Shouldn't the deadline go into the least significant bits? @edwardalee?

**Review by lhstrh — APPROVED** [review]
I've made a pass over this PR, which is, needless to say, quite enormous. It represents an impressive amount of work and brings many design improvements -- excellent work, @Soroosh129! 

Some things I noticed while browsing the code: 
 - There is some code replication across the different schedulers; it would be better to have some reusable default implementations and let the C compiler optimize out the extra function calls.
 - The entanglement of code for federated execution worries me, and it seems that not all schedulers offer support for federation. While the `#ifdef` approach was a good first start, I think it would be better to develop a proper API instead.
 - Each scheduler would benefit from an overall description of its architecture at the beginning of the file. Without such description, it is really quite difficult to figure out what's going on.

None of these issues, however, should block the PR from being merged.

**Comment by Soroosh129** [comment]
Thanks @lhstrh for your review. Your comments are very helpful.

> There is some code replication across the different schedulers; it would be better to have some reusable default implementations and let the C compiler optimize out the extra function calls.

Would you please point to specific duplications that you found, so I know where I should pay attention to? I tried my best to factor all the common code I could find (that made sense to me to factor out) into common files (`scheduler_instance.h` and `scheduler_sync_tag_advance.c`), so I would appreciate pointers to places that I missed if you have them. 

> it seems that not all schedulers offer support for federation

They do. `scheduler_NP` is special because it operates lock-free. 
However, in federated execution, a reaction could be triggered by `federate.c` at the _current_ executing reaction level (i.e., the level of reactions that are currently executing), causing a race condition between worker threads that are accessing the reaction queue for the current level (to get a reaction to execute) and federate.c (that is trying to trigger a reaction at the current level). To avoid this race condition, a mutex is locked to access the reaction queue of the current reaction level. This obviously complicates the code of `scheduler_NP` slightly

> While the #ifdef approach was a good first start, I think it would be better to develop a proper API instead.

I would appreciate it if you could give me a concrete suggestion on what this proper API would look like and how to use it here. Are we talking about something like `isFederated()` or something else?

> Each scheduler would benefit from an overall description of its architecture at the beginning of the file. Without such description, it is really quite difficult to figure out what's going on.

You're right. I don't have the bandwidth to write them now, but I agree that these descriptions are definitely required to properly convey the design of these schedulers to people.

File: core/threaded/scheduler_GEDF_NP_CI.c
Comment (Soroosh129) [reply core/threaded/scheduler_GEDF_NP_CI.c:439]: I agree that we should document this in an issue. I didn't get to thoroughly investigate this, but we should give this issue a closer look. In the previous design, calling `pqueue_free(reaction_q)` would cause the same type of "double free" errors if I remember correctly.

Comment (lhstrh) [reply core/threaded/scheduler_GEDF_NP_CI.c:439]: OK, I created an issue for you: https://github.com/lf-lang/reactor-c/issues/41

File: core/threaded/scheduler_GEDF_NP.c
Comment (lhstrh) [reply core/threaded/scheduler_GEDF_NP.c:33]: https://github.com/lf-lang/reactor-c/issues/42

**Comment by lhstrh** [comment]
> Would you please point to specific duplications that you found, so I know where I should pay attention to? I tried my best to factor all the common code I could find (that made sense to me to factor out) into common files (`scheduler_instance.h` and `scheduler_sync_tag_advance.c`), so I would appreciate pointers to places that I missed if you have them.

I would have to go back into review mode and start digging. Let's punt on this for now; whatever code duplication we encounter in the future we can address by introducing default implementation.
 
> > it seems that not all schedulers offer support for federation
> 
> They do. `scheduler_NP` is special because it operates lock-free. However, in federated execution, a reaction could be triggered by `federate.c` at the _current_ executing reaction level (i.e., the level of reactions that are currently executing), causing a race condition between worker threads that are accessing the reaction queue for the current level (to get a reaction to execute) and federate.c (that is trying to trigger a reaction at the current level). To avoid this race condition, a mutex is locked to access the reaction queue of the current reaction level. This obviously complicates the code of `scheduler_NP` slightly.

Thanks for clarifying, @Soroosh129!
 
> > While the #ifdef approach was a good first start, I think it would be better to develop a proper API instead.
> 
> I would appreciate it if you could give me a concrete suggestion on what this proper API would look like and how to use it here. Are we talking about something like `isFederated()` or something else?

A simple improvement that I was thinking of was to take whatever code is currently surrounded by `#ifdef` `#endif` and put that code in a function (along with the preprocessor commands). My hope is that this would promote readability, but we can only find out by trying. I created an issue: https://github.com/lf-lang/reactor-c/issues/43.

> > Each scheduler would benefit from an overall description of its architecture at the beginning of the file. Without such description, it is really quite difficult to figure out what's going on.
> 
> You're right. I don't have the bandwidth to write them now, but I agree that these descriptions are definitely required to properly convey the design of these schedulers to people.

Issue created: https://github.com/lf-lang/reactor-c/issues/42

File: core/threaded/scheduler_GEDF_NP_CI.c
Comment (edwardalee) [reply core/threaded/scheduler_GEDF_NP_CI.c:406]: I think this relies on the fact that the deadline is inherited by all upstream reactions. It means that an entire chain of reactions with deadlines can execute before a parallel chain without deadlines.

**Pull request merged**
